design-documents/allowed-http-methods-for-actions.md

```diff
@@ -0,0 +1,56 @@
+### Overview
+
+To improve security and logistics we need to allow limiting Actions to processing only requests with certain HTTP methods and add those limitations to as many existing Actions as possible.
+There are many vulnerabilities caused by actions processing both GET and POST requests and thus allowing bypassing security validations like form key validation. Also limiting actions to processing only requests with certain methods would serve as self-documentation for Action classes and improve consistency of server side for client code and functional tests.
+
+### Theory of Operation
+
+Since we use by-convention routing it would be inconsistent to defined actions’ allowed HTTP methods in a config file, so it’s better to create marker interfaces like these:
+
+>interface HttpGetActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPostActionInterface extends ActionInterface
+>
+>{
+>
+>}
+>
+>
+>interface HttpPutActionInterface extends ActionInterface
+>
+>{
+>
+>}
+
+etc.
+
+One interface per each HTTP method (using list of HTTP methods from \Zend\Http\Request::METHOD_* constants). We will have an exentable map *interface* => *HTTP method* passed to the validator to allow custom HTTP methods to be processed.
+
+Request method will be validated in the FrontController before executing a matched action, somewhat like this:
+
+>foreach ($interfaceMap as $interface => $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;if ($actionInstance instanceof $interface && $request->getMethod() !== $httpMethod) {
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;throw new NotFoundException(__(‘Page not found.’));
+>
+>&nbsp;&nbsp;&nbsp;&nbsp;}
+>
+>}
+
+ 
+To limit already existing Actions and get developers to know these new interfaces a temporary event listener will be added for controller_action_postdispatch event to log HTTP methods used with actions and then the full suite of functional tests (MTF and MFTF) would be ran to get information on as much actions as possible.
```

Comment:
Can we leave logging in place to let extension developers know that they need to update their extensions?

Reply:
I remember we discussed this, I just forgot to mention it - we can add a debug message each time an action without HTTP method declared s processed

Reply:
I'm ok with not having this if we going to have a mess detector rule.